src/Spago/Config.purs

```diff
@@ -174,130 +175,212 @@ type ReadWorkspaceOptions =
+        , compatibleCompiler: compiler
+        , backend: workspace.config.backend
+        , buildOptions:
+            { output: workspace.config.buildOpts >>= _.output <#> \o -> withForwardSlashes $ rootPath </> o
+            , censorLibWarnings: _.censorLibraryWarnings =<< workspace.config.buildOpts
+            , statVerbosity: _.statVerbosity =<< workspace.config.buildOpts
+            }
+        , doc: Just workspace.doc
+        , workspaceConfig: workspace.config
+        , rootPackage: workspace.rootPackage
+        }
+    }
+  where
+  readConfig' :: ∀ s. _ -> State.StateT s _ _
+  readConfig' = State.lift <<< readConfig
+
+  walkDirectoriesUpFrom dir = do
+    maybeConfig <- tryReadConfigAt configFile
+
+    for_ maybeConfig \config ->
+      for_ config.yaml.package \package ->
+        -- If there is a package in this directory, remember it
+        State.modify_ \s -> s
+          { loadedPackages = Map.insert dir { package, config } s.loadedPackages
+          , closestPackage = s.closestPackage <|> Just package.name
           }
 
-  { right: otherPackages, left: failedPackages } <- partitionMap identity <$> traverse readWorkspaceConfig otherConfigPaths
-  unless (Array.null failedPackages) do
-    logWarn $ [ toDoc "Failed to read some configs:" ] <> failedPackages
+    whenM (FS.exists $ dir </> "spago.yml") $
+      State.modify_ \s -> s { misnamedConfigs = Array.cons dir s.misnamedConfigs }
+
+    case maybeConfig of
+      Just { doc, yaml: { workspace: Just workspace, package } } -> do
+        -- Finally, found the "workspace" config!
+        rootPath <- Path.mkRoot dir
+        pure { workspace: { config: workspace, doc, rootPackage: package }, rootPath }
+      _ -> do
+        -- No workspace in this directory => recur to parent directory (unless it's already root)
+        when (parentDir == dir) $
+          dieForLackOfSpagoYaml
+        walkDirectoriesUpFrom parentDir
+
+    where
+    configFile = dir </> spagoYaml
+    parentDir = Path.dirname dir
+
+  loadSubprojectConfigs rootPath = do
+    candidates <- liftAff $ Glob.gitignoringGlob
+      { root: rootPath
+      , includePatterns: [ "**/" <> spagoYaml ]
+      , ignorePatterns: [ "**/node_modules/**", "**/.spago/**" ]
+      }
 
-  -- We prune any configs that use a different workspace.
-  -- For reasoning, see https://github.com/purescript/spago/issues/951
-  let configPathsWithWorkspaces = otherPackages # Array.mapMaybe \readResult -> readResult.packagePath <$ readResult.config.yaml.workspace
-  unless (Array.null configPathsWithWorkspaces) do
-    logDebug $ "Found these paths with workspaces: " <> String.joinWith ", " (Path.quote <$> configPathsWithWorkspaces)
+    -- Traversing directories (not files) and doing it in sorted order ensures
+    -- that parent directories come before their subdirectories. That way we
+    -- can remember workspaces that we find along the way and avoid trying to
+    -- load their subprojects that come later.
```

Comment:
We should still mention https://github.com/purescript/spago/issues/951 somewhere in this area.

Reply:
I don't understand this one. Mention how?

Reply:
There currently is a comment that links to that one - I think it's useful to leave a paper trail if someone gets in the code and comes up with the question "why are we excluding some packages"

Reply:
Added the previously existing comment. Also added a mention of discussion in https://github.com/purescript/spago/pull/1316#issuecomment-2629011927